mitiq/zne/inference.py

```diff
@@ -397,12 +378,183 @@ def is_converged(self) -> bool:
+            scale_noise: Noise scaling function.
+            num_to_average: Number of times to call scale_noise at each scale
+                factor.
+        """
+        self.reset()
+        self._batch_populate_instack()
+        kwargs = self._get_keyword_args()
+
+        # Get all noise-scaled circuits to run
+        to_run = self._generate_circuits(qp, scale_noise, num_to_average)
+
+        # Run the circuits in a batch
+        # TODO: Deal with keyword args better.
+        res = batched_executor(to_run, kwargs=kwargs)
+
+        # Average the expectation results
+        self._outstack = [
+            np.average(res[i * num_to_average: (i + 1) * num_to_average])
+            for i in range(len(res) // num_to_average)
+        ]
+
+        return self
+
+    def run(
+        self,
+        qp: QPROGRAM,
+        executor: Tuple[Callable[..., float], Callable[..., List[float]]],
+        scale_noise: Callable[[QPROGRAM, float], QPROGRAM],
+        num_to_average: int = 1,
+    ) -> "BatchedFactory":
+        """Computes the expectation values at each scale factor and stores them
+        in the factory. If the executor returns a single expectation value, the
+        circuits are run sequentially. If the executor is batched and returns
+        a list of expectation values (one for each circuit), then the circuits
+        are sent to the backend as a single job. To detect if an executor is
+        batched, it must be annotated.
+
+        Args:
+            qp: Quantum circuit to run.
+            executor: A "single executor" (1) or a "batched executor" (2).
+                (1) A function which inputs a single circuit and outputs a
+                single expectation value of interest.
+                (2) A function which inputs a list of circuits and outputs a
+                list of expectation values (one for each circuit).
+            scale_noise: Noise scaling function.
+            num_to_average: Number of times to call scale_noise at each scale
+                factor.
+        """
+        self.reset()
+        executor_annotation = inspect.getfullargspec(executor).annotations
+        if executor_annotation.get("return") in (
+            List[float],
+            Sequence[float],
+            Tuple[float],
+            Iterable[float],
+        ):
+            self.run_batched(qp, executor, scale_noise, num_to_average)
+            return self
+
+        def _noise_to_expval(scale_factor, **exec_params) -> float:
+            """Evaluates the quantum expectation value for a given
+            scale_factor and other executor parameters."""
+            expectation_values = []
+            for _ in range(num_to_average):
+                scaled_qp = scale_noise(qp, scale_factor)
+                expectation_values.append(executor(scaled_qp, **exec_params))
+            return np.average(expectation_values)
+
+        return self.iterate(
+            _noise_to_expval, max_iterations=len(self._scale_factors) + 1
```

Comment:
Afaict this is identical to `AdaptiveFactory.run()`, but with `max_iterations` set to `len(self._scale_factors) + 1`. As such, if you follow @andreamari's advice (that I talk about in the previous comment above), this can be replaced by a call to `self.run` with the appropriate args, thus removing the "code smell".

Reply:
Yes it is identical. I'm proposing to not have a common run method between `BatchedFactory` and `AdaptiveFactory` (see below), so this will eventually become outdated.